pkg/server/node_auth.go

```diff
@@ -0,0 +1,130 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"fmt"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/security"
+	"istio.io/istio/pkg/spiffe"
+	"istio.io/istio/pkg/util/sets"
+	"istio.io/istio/security/pkg/server/ca"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ClusterNodeAuthorizer is a component that implements a subset of Kubernetes Node Authorization
+// (https://kubernetes.io/docs/reference/access-authn-authz/node/) for Istio CA within one cluster.
+// Specifically, it validates that a node proxy which requests certificates for workloads on its
+// own node is requesting valid identities which run on that node (rather than arbitrary ones).
+// This implementation is based on Istio, but ensures the pod informer is synced before
+// creating the index.
+// (https://github.com/istio/istio/blob/1.22.1/security/pkg/server/ca/node_auth.go#L74)
+
+type ClusterNodeAuthorizer struct {
+	trustedNodeAccounts sets.Set[types.NamespacedName]
+	pods                kclient.Client[*v1.Pod]
+	nodeIndex           kclient.Index[ca.SaNode, *v1.Pod]
+}
+
+func NewClusterNodeAuthorizer(client kube.Client, trustedNodeAccounts sets.Set[types.NamespacedName]) *ClusterNodeAuthorizer {
+	pods := kclient.NewFiltered[*v1.Pod](client, kclient.Filter{
+		ObjectFilter:    client.ObjectFilter(),
+		ObjectTransform: kube.StripPodUnusedFields,
+	})
+
+	stopCh := make(chan struct{})
+	client.RunAndWait(stopCh)
+	kube.WaitForCacheSync("nodeAuth", stopCh, pods.HasSynced)
+
+	// Add an Index on the pods, storing the service account and node. This allows us to later efficiently query.
+	index := kclient.CreateIndex[ca.SaNode, *v1.Pod](pods, func(pod *v1.Pod) []ca.SaNode {
+		if len(pod.Spec.NodeName) == 0 {
+			return nil
+		}
+		if len(pod.Spec.ServiceAccountName) == 0 {
+			return nil
+		}
+		return []ca.SaNode{{
+			ServiceAccount: types.NamespacedName{
+				Namespace: pod.Namespace,
+				Name:      pod.Spec.ServiceAccountName,
+			},
+			Node: pod.Spec.NodeName,
+		}}
+	})
+
+	return &ClusterNodeAuthorizer{
+		pods:                pods,
+		nodeIndex:           index,
+		trustedNodeAccounts: trustedNodeAccounts,
+	}
+}
+
+// authenticateImpersonation will verify the caller is authorized to impersonate the requested identity
+func (na *ClusterNodeAuthorizer) authenticateImpersonation(caller security.KubernetesInfo, requestedIdentityString string) error {
+	callerSa := types.NamespacedName{
+		Namespace: caller.PodNamespace,
+		Name:      caller.PodServiceAccount,
+	}
+
+	// First, make sure the caller is allowed to impersonate, in general
+	if _, f := na.trustedNodeAccounts[callerSa]; !f {
+		return fmt.Errorf("caller (%v) is not allowed to impersonate", caller)
+	}
+
+	// // Next, make sure the identity they want to impersonate is valid, in general
```

Comment:
turbo nitpick (not a blocker): There are a few comments here that have `// //`, which I guess is an artifact from copy-pasting?